lib/HLSL/DxilContainerReflection.cpp

```diff
@@ -285,6 +287,8 @@ HRESULT CreateDxilShaderReflection(const DxilProgramHeader *pProgramHeader, cons
 HRESULT CreateDxilLibraryReflection(const DxilProgramHeader *pProgramHeader, const DxilPartHeader *pRDATPart, REFIID iid, void **ppvObject) {
   if (!ppvObject)
     return E_INVALIDARG;
+  if (!IsEqualIID(__uuidof(ID3D12LibraryReflection), iid))
+    return E_INVALIDARG;
```

Comment:
E_NOINTERFACE would be the usual error to return from factory functions like this where the object being created doesn't support the requested interface.

Reply:
Thanks for this.  As it turns out, I realized that `IUnknown` should be allowed for these create functions as well.  There's regression risk for disallowing them, and we know whether it's a shader or library from the blob, so we don't need it for that.  For shader reflection, if `IUnknown` is used on create, the version defaults to the latest (D3D12).

Reply:
The way I usually implement this sort of thing is to create the object and then use QueryInterface to return the appropriate interface from it.  That way we don't need to update it if the set of interfaces supported by the object changes.

Reply:
That's what we normally do, but for shader reflection, we have some code to capture the initial interface version you use to initialize the reflection object versioning.  This is so it behaves as expected for the interface version.  You're then not allowed to QI to a different interface version, since that interface would misbehave, since we don't support multiple interfaces of different versions on the same underlying reflection object.  There would be another way to implement this, like with different interface implementations for each version, delegating most things to a central implementation and adapting for interface differences, but that's more complicated, and not the way this was implemented.  Nor is it likely that someone using one version of the reflection interface is going to turn around and use another version of the reflection interface through QI.  These normally would have been defined based on the SDK you linked with.  So, there is a check for the interface used before even creating the reflection, skipping it altogether if it's not valid.

This pattern is carried over to library reflection creation, where there's only one interface version.  It could be ripped out, and that would change it to creating the reflection, then releasing it if you provided the wrong interface.  However, if we version the library reflection interface to add some needed properties (not unlikely given current requests for work graphs), we might need a similar mechanism to the shader reflection versioning, and thus a similar sequence of checking the version before creating the reflection object.

Reply:
It'd be great to capture some of this in the code because it is extremely unusual and not at all obvious from reading it.